generator/lib/src/realm_model_info.dart

```diff
@@ -87,6 +87,10 @@ class RealmModelInfo {
       yield 'Stream<RealmObjectChanges<$name>> get changes => RealmObject.getChanges<$name>(this);';
       yield '';
 
+      yield '@override';
+      yield '$name freeze() => RealmObject.freezeObject<$name>(this);';
+      yield '';
```

Comment:
The reason why we want to override this in the generated class is to specialize the returned value to match the object type.

Reply:
I see a two options here to consider:
* `void freeze()` swizzle the existing realm and handle to frozen versions. Ie. change the object in place.
* `X get frozen` use property style (matches better with `dynamic`) and a bit shorter

Any thoughts on pros & cons?

Reply:
I don't like the property syntax as it looks misleadingly cheap. Not sure if that's the case for dart, but when I look at properties, I don't expect some complicated logic to be happening when I invoke them. It's not like freezing is super expensive, but it does a fair amount of allocations and I'd like this to be more explicit. It's not like we have the frozen version of the object lying around and you just access it through the property - you're performing some operation which is better conveyed with a method.

Regarding freezing in place - I mildly dislike it because it's less flexible. Let's say you have a method that accepts a RealmObject as an argument. You may want to obtain a frozen version of that object, but if it's an in-place operation, you'll modify your inputs and make the object unusable for the caller. Let's say you have data-bound an object to your UI, if you pass it to some method that freezes it, your databinding becomes unusable.

Reply:
As I understand freezing is super cheap for core (@finnschiermer convinced me so in a video once 😄 ). Basically it has no upfront cost - like O(1) with extremely low constant factor. But it may make it difficult to reap storage from old versions later.

If we froze in-place it would literally just be:
```dart
    final frozenRealm = object.realm.freeze();
    final frozenHandle = realmCore.resolveObject(object, frozenRealm)!;
``` 
I think `freeze()` reads like a `void` method that changes the object in place. So, if we go with `freeze` I would prefer that semantic. Even though you need to create a new object if we go with the property `obj.frozen` then it is no more expensive than say indexing into a realm list.

Reply:
I think Finn may be overselling the cheapness here. It may be super cheap for Core, but it's noticeably expensive for us, particularly when we don't have a cached frozen Realm at the version that we're at. You can see the amount of work that goes into `Realm::get_frozen_realm` both in terms of validations, but also copying structures around. While I haven't measured it for dart, I remember it was many times more expensive than resolving a link for .NET.

Re: freezing in place - I believe all SDKs are using the `freeze()` terminology and returning a new instance rather than freezing an existing instance in place. While we could use `getFrozen()` as an alternative, I'd rather not deviate from the terminology already adopted.

Reply:
```c++
SharedRealm Realm::get_frozen_realm(Config config, VersionID version)
{
    auto coordinator = RealmCoordinator::get_coordinator(config.path);
    SharedRealm realm = coordinator->get_realm(std::move(config), util::Optional<VersionID>(version));
    realm->set_auto_refresh(false);
    return realm;
}
```
don't smell expensive to me, but I would need to investigate more, and measure. 

Anyway, the argument that both kotlin, swift and dotnet adopted `T freeze()` is a winning one, but it leads me to another question. Should we capture in the type system that the object is frozen, ie. return a `FrozenT` instead of a `T`?

Reply:
I think that may be a bit too annoying as we'd need to essentially shadow all our types + generate frozen version for the user types. This also means that a method accepting/being constrained on `RealmObject` won't work with `FrozenObject`. I'm open to being convinced otherwise, but I feel it's a fair amount of work that may end up being more confusing than helpful.

Reply:
It would be significant work yes, but I'm a big fan of capturing constraints in the type system. However, let us land this PR separately, and discuss later.